### PR TITLE
Add __table function and unit tests

### PR DESCRIPTION
**Pull request by annie**
This PR ports over @mbostock's `__table` function, as defined in https://github.com/observablehq/observablehq/pull/8300, and adds missing support for the `lt` and `gt` filter operations. Also adds unit tests.

### CODE REVIEW

File: test/table-test.mjs [hunk 1]
```diff
@@ -35,178 +35,261 @@ const baseOperations = {
...
-    }
-  ];
-
-  invalidFilters.map((filter) => {
-    const operations = {
-      ...baseOperations,
-      filter: [filter]
-    };
-    assert.throws(() => makeQueryTemplate(operations, source), /Invalid filter operation/);
+describe("makeQueryTemplate", () => {
```
Comment (annie) [inline test/table-test.mjs:38]: the only change i made here was to wrap the `makeQueryTemplate` tests in a `describe` – if you check "Hide whitespace", this diff will look much better :)

Comment (annie) [inline test/table-test.mjs:38]: the only change i made here was to wrap the `makeQueryTemplate` tests in a `describe` – if you check "Hide whitespace", this diff will look much better 😄

File: src/table.mjs [hunk 2]
```diff
@@ -1,7 +1,27 @@
...
+  );
+}
+
 export const __query = Object.assign(
   // This function is used by table cells.
   async (source, operations, invalidation) => {
-    const args = makeQueryTemplate(operations, await source);
+    // For cells whose data source is an in-memory table, we use JavaScript to
+    // apply the table cell operations, instead of composing a SQL query.
+    // TODO Do we need to port over all the logic from canDisplayTable, or is
```
Comment (annie) [inline src/table.mjs:21]: [canDisplayTable](https://github.com/observablehq/observablehq/blob/main/notebook-next/src/worker/display.js#L24) contains a lot of logic, and i don't know if we need to copy that over to stdlib? it seems like checking the general structure of `source` is sufficient here, since the source needs to have passed the `canDisplayTable` check to even show up as a data source option in the first place.

File: src/table.mjs [hunk 3]
```diff
@@ -1,7 +1,27 @@
...
+    // this sufficient?
+    if (Array.isArray(source) || isTypedArray(source)) return __table(source, operations);
+    const args = makeQueryTemplate(await source, operations);
```
Comment (mbostock) [inline src/table.mjs:24]: At this point, we could throw an error if the source is not a database client. We have an isDatabaseClient function somewhere that does the duck test.

File: src/table.mjs [hunk 4]
```diff
@@ -1,7 +1,27 @@
...
+export function isTypedArray(value) {
+  return (
+    value instanceof Int8Array ||
+    value instanceof Int16Array ||
+    value instanceof Int32Array ||
+    value instanceof Uint8Array ||
+    value instanceof Uint8ClampedArray ||
+    value instanceof Uint16Array ||
+    value instanceof Uint32Array ||
+    value instanceof Float32Array
```
Comment (mbostock) [inline src/table.mjs:12]: TODO Float64Array?

File: src/table.mjs [hunk 5]
```diff
@@ -73,7 +93,7 @@ async function* accumulateQuery(queryRequest) {
  * of sub-strings and params are the parameter values to be inserted between each
  * sub-string.
  */
- export function makeQueryTemplate(operations, source) {
+ export function makeQueryTemplate(source, operations) {
```
Comment (annie) [inline src/table.mjs:96]: note to self: revert this change because `makeQueryTemplate` is exported, so making this change would require a major version bump

File: src/table.mjs (hunk 3 above)
Comment (annie) [reply src/table.mjs]: done

File: src/table.mjs (hunk 4 above)
Comment (annie) [reply src/table.mjs]: done 🤞

File: src/table.mjs [hunk 6]
```diff
@@ -231,3 +271,91 @@ function likeOperand(operand) {
...
+      }
+      default:
+        throw new Error(`unknown filter type: ${type}`);
+    }
+  }
+  for (const {column, direction} of reverse(operations.sort)) {
+    const compare = direction === "desc" ? descending : ascending;
+    source.sort((a, b) => compare(a[column], b[column]));
+  }
+  if (operations.slice) {
```
Comment (mbostock) [inline src/table.mjs:341]: According to our schema, operations.slice is always present, but we only need to slice if operations.slice.from or operations.slice.to is non-null (or non-zero and non-infinity, if preferred).

File: src/table.mjs [hunk 7]
```diff
@@ -231,3 +271,91 @@ function likeOperand(operand) {
...
+    source = source.slice(
+      operations.slice.from ?? 0,
+      operations.slice.to ?? Infinity
+    );
+  }
+  if (operations.select?.columns) {
```
Comment (mbostock) [inline src/table.mjs:347]: And likewise here, options.select is required, so we don’t need the `?`.

File: src/table.mjs [hunk 8]
```diff
@@ -231,3 +271,91 @@ function likeOperand(operand) {
...
+      schema = operations.select.columns.map((c) => schemaByName.get(c));
+    } else if (columns) {
+      columns = operations.select.columns;
+    }
+    source = source.map((d) =>
+      Object.fromEntries(operations.select.columns.map((c) => [c, d[c]]))
+    );
+  }
+  if (schema) source.schema = schema;
+  else if (columns) source.columns = columns;
```
Comment (mbostock) [inline src/table.mjs:359]: We should move this assignment to within the if (operations.select.columns) block; otherwise, we should allow the source’s existing schema and columns to pass through unaffected, and we shouldn’t mutate the passed-in source (i.e., we only want to assign these properties after creating a copy of source by calling source.map).

File: src/table.mjs [hunk 9]
```diff
@@ -1,6 +1,46 @@
...
+    (typeof value.sql === "function" ||
+      (typeof value.queryTag === "function" &&
+        (typeof value.query === "function" ||
+          typeof value.queryStream === "function"))) &&
+    (mode !== "table" || typeof value.describeColumns === "function") &&
+    value !== __query // don’t match our internal helper
+  );
+}
+
+export function isTypedArray(value) {
```
Comment (mbostock) [inline src/table.mjs:23]: This does not need to be exported.

```suggestion
function isTypedArray(value) {
```

File: src/table.mjs [hunk 10]
```diff
@@ -231,3 +271,91 @@ function likeOperand(operand) {
   return {...operand, value: `%${operand.value}%`};
 }
 
+// This function applies table cell operations to an in-memory table (array of objects).
+export function __table(source, operations) {
+  let {schema, columns} = source;
+  for (const {type, operands} of operations.filter) {
+    const column = operands.find(({type}) => type === "column").value;
```
Comment (mbostock) [inline src/table.mjs:278]: According to our schema, the column operand must always be listed first. So, in theory we could write it faster like this:

```suggestion
    const [{value: column}] = operands;
```

File: src/table.mjs [hunk 11]
```diff
@@ -231,3 +271,91 @@ function likeOperand(operand) {
...
+    const resolved = operands.filter(({type}) => type === "resolved");
```
Comment (mbostock) [inline src/table.mjs:279]: I’m not sure this is right. According to our schema, TableFilterOperands can have the following type:

* _primitive_
* _date_
* _cell_

I don’t see us using the type _resolved_ anywhere. I’m also not sure what the type _cell_ means… I’m guessing that was speculative and should be removed from the schema. Anyway, I think we can say this:

```suggestion
    const values = operands.slice(1).map(({value}) => value);
```

File: src/table.mjs [hunk 12]
```diff
@@ -231,3 +271,91 @@ function likeOperand(operand) {
...
+        source = source.filter((d) => d[column] > value);
+        break;
+      }
+      default:
+        throw new Error(`unknown filter type: ${type}`);
+    }
+  }
+  for (const {column, direction} of reverse(operations.sort)) {
+    const compare = direction === "desc" ? descending : ascending;
+    source.sort((a, b) => compare(a[column], b[column]));
```
Comment (mbostock) [inline src/table.mjs:339]: This might be mutating the input _source_. We need to make sure we make a defensive copy before sorting (but ideally we avoid unnecessary defensive copies, too).

**Review by mbostock — CHANGES_REQUESTED** [review]
If I have time tonight I’ll put up a PR to this PR to help this along! Thanks Annie!

File: src/table.mjs (hunk 11 above)
Comment (mbostock) [reply src/table.mjs:279]: Ah ha. The compiler replaces _primitive_, _date_, and _cell_ filter operands with _resolved_. So it was correct as-is. I’ll update my PR accordingly.

File: src/table.mjs [hunk 13]
```diff
@@ -231,3 +271,91 @@ function likeOperand(operand) {
...
+// This function applies table cell operations to an in-memory table (array of objects).
+export function __table(source, operations) {
+  let {schema, columns} = source;
+  for (const {type, operands} of operations.filter) {
+    const column = operands.find(({type}) => type === "column").value;
+    const resolved = operands.filter(({type}) => type === "resolved");
+    switch (type) {
+      case "eq": {
+        const [{value}] = resolved;
+        source = source.filter((d) => d[column] === value);
```
Comment (mbostock) [inline src/table.mjs:283]: We’re using strict equality here, but the resolved value will always be a string (in our UI), whereas the type of the column might be a number. I think this means we’ll have to use loose equality here instead… and I’m not sure how we’ll handle dates.

**Review by libbey-observable — APPROVED** [review]

**Review by mbostock — APPROVED** [review]

File: src/table.mjs [hunk 14]
```diff
@@ -1,12 +1,50 @@
...
 export const __query = Object.assign(
-  // This function is used by table cells.
   async (source, operations, invalidation) => {
-    const args = makeQueryTemplate(operations, await source);
-    if (!args) return null; // the empty state
-    return evaluateQuery(await source, args, invalidation);
+    source = await source;
+    // For cells whose data source is an in-memory table, we use JavaScript to
+    // apply the table cell operations, instead of composing a SQL query.
+    if (Array.isArray(source) || isTypedArray(source)) return __table(source, operations);
```
Comment (mbostock) [inline src/table.mjs:43]: The isTypedArray check here isn’t enough—we also need to promote the typed array into an array of {value} objects so that __table can work as expected.

File: src/table.mjs (hunk 2 above)
Comment (mbostock) [reply src/table.mjs]: I’ve ported this logic over so that we can show the correct errors and also handle promotion of primitive arrays.
